[1단계 - 자동차 경주 구현] 다온(조형익) 미션 제출합니다.

ikjo39 opened the pull request:
안녕하세요 리뷰어님 6기 다온입니다.

이번 자동차 경주 미션은 @테드와 페어로 진행한 미션입니다.

효율적인 리뷰를 위해 아래 참고사항을 정리해보았습니다.

## 미션을 수행하며 진행된 일

돌아가는 쓰레기 코드부터 만들고 리팩토링 해보자란 마음으로 시작하게 되었는데요.

페어 프로그래밍을 처음이라 익숙해지는데 조금 시간이 들어 리팩토링이 잘 되었는지 확신이 부족합니다.

---

## 코드 작성시 신경 쓴 점

- MVC 패턴을 적용해보았습니다.
    - UI와 비즈니스 로직을 분리하기 위해 MVC 패턴을 채택하여 적용하였습니다.
- 난수를 위한 테스트 코드에 특히 신경썼어요.
    - 우선 인터페이스를 활용하여 비즈니스 로직과 테스트 코드에서 난수 로직이 다르게 작동하도록 구현하였습니다.
        - 테스트 코드내에서는 인터페이스를 활용한 `MovingStub`으로 각 자동차의 이동을 커스텀했습니다.
- 자동차 경주 결과를 저장하기 위해 `LinkedHashMap`을 이용하였습니다.

---

## 질문

- view 에서 입력값 변환의 책임은 명확한 기준이 있는지 컨벤션을 정해야하는지 궁금합니다.
    - 페어와 아래 3가지를 view layer 에서 해도 될지 토론하였지만 결론이 나지 않았습니다.
        - 타입변환에 대한 예외처리
        - 구분자로 split 후 `List<>` 변환
        - 객체 포장
- `OutputView` 내 `printResult()` 메서드에서 getter를 사용하여 출력 메세지를 만들었는데요.
    - View 에서 비즈니스 로직을 다루지 않는다는 MVC 패턴 규칙을 위배하지 않았나 궁금합니다.

이상입니다!

Kimprodp committed:
docs: 구현할 기능 목록 추가

Kimprodp committed:
feat: 자동차 이름을 입력받아 자동차를 생성한다.

Kimprodp committed:
feat: 이동횟수를 입력받아 레이싱 게임을 생성한다.

Kimprodp committed:
feat: 레이싱 게임 구현

Kimprodp committed:
refactor: MVC패턴 적용

Kimprodp committed:
refactor: 매직넘버 상수화 및 의미있는 메서드(변수)명으로 수정

Kimprodp committed:
refactor: 반환 타입 변경 및 일부 메서드명 수정

- Cars 내 makeCasMove 메서드 반환타입 변경 LinkedHashMap -> RoundResult

Kimprodp committed:
refactor: 반환 타입 변경

- RacingGame 내 run 메서드 반환타입 변경 List<RoundResult> -> TotalResult

Kimprodp committed:
test: RoundResultTest 및 TotalResultTest 추가

sihyung92 commented on src/test/java/racingcar/model/CarsTest.java:
```diff
@@ -0,0 +1,48 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.LinkedHashMap;
```
사용하지 않는 임포트가 있어용!
한번씩 코드 작성 중에 [import 정리](https://atoz-develop.tistory.com/entry/IntelliJ-%EC%9C%A0%EC%9A%A9%ED%95%9C-%EB%8B%A8%EC%B6%95%ED%82%A4-import%EB%AC%B8-%EC%A0%95%EB%A6%AC%ED%95%98%EA%B8%B0) 눌러주기~~

안 쓰는 import 때문에 추후에 이걸 정리하게 되면 다른 사람이 기능 추가한 코드와 컨플릭트가 생기는 경우가 종종 있습니다

ikjo39 replied:
9b9221e8ad43a92e607d14e978b08ebc51883f7e 리팩토링 진행 후 테스트 코드 정리를 놓친거 같아요:sob:
커밋하기전 다시 한번 체크해봐야겠어요 🔥

sihyung92 commented on docs/README.md:
```diff
@@ -0,0 +1,14 @@
+## 필수 기능
```
<커밋 관련 리뷰는 반영할 필요는 없고 다음부터 참고해주세요>
d147bec4 해쉬의 커밋명은 '입력횟수'가 맞을 것 같습니다!
b0c92636 해쉬는 numberGenerator와 관련한 리뷰를 하기 위해 해당 해쉬를 찾았는데, 너무 큰 기능이 한 커밋에 담겨있어서 어떤 생각을 가지고 만드셨는지 짐작이 어려웠습니다. 커밋 단위를 잘게 쪼개보시면 어떨까요?

ikjo39 replied:
A1. 엇 기능 요구사항에서는 입력횟수라 정의했었네요! 놓쳤던 부분이에요 감사합니다 😄 
A2. 웨지 의견에 완전 동의해요. 요구사항을 빠르게 개발하자란 욕심에 '괜찮겠지..'하며 넘겼던게 생각나 얼굴이 붉어지네요.

sihyung92 commented on src/main/java/racingcar/generator/NumberGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar.generator;
+
+public interface NumberGenerator {
```
왜 인터페이스로 추상화 하셨나요?
랜덤 숫자와 스텁용 두 가지 활용처 밖에 없는데 과도한 추상화 같아요

무언가를 추상화 할 땐 이 부분을 나중에 내가 절대 못 고치게 될까? / 변화가 확실한 코드일까? 를 고려해서 선택하시면 좋아요

내가 언제든 고칠 수 있는 코드면 추상화하지 않는게 낫습니다.

그런데 500명이 내 코드에 의존하고 있어서 내맘대로 수정배포할 수 없는 상황이 올 거라면 변경이 될만한 부분들을 추상화해서 구현체만 갈아낄 수 있도록 하면 좋습니다.

프로덕션 기능 외에 테스트를 위해 추상화하는 건 좋지않은 선택입니다. 프로덕션 코드 끼리도 복잡도가 높은데, 테스트의 의존성까지 생기면 더 힘들어집니다

RandomNumberGenerator 외에 뭔가 새로운 요구사항이 들어올 수 있을거 같다~ 판단 해주시고 추상화 제거 리뷰 반영은 선택적으로 해주세요

ikjo39 replied:
> 프로덕션 기능 외에 테스트를 위해 추상화하는 건 좋지않은 선택입니다. 프로덕션 코드 끼리도 복잡도가 높은데, 테스트의 의존성까지 생기면 더 힘들어집니다

우선 인터페이스는 웨지가 조금은 예상하셨겠지만 랜덤 숫자를 테스트하기 위해 사용하였습니다.
프리코스 때 학습하여 랜덤을 테스트할 때는 이 방법 하나만 있는 줄 알고 사용하게 되었어요.
생각해보지 못했던 관점이라 머리를 한 대 맞은 기분이에요 😅 

추상화 제거 후 랜덤 숫자 테스트는 한번 열심히 고안해볼게요!

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,42 @@
+            position++;
+        }
+        return position;
+    }
```
제가 좋아하는 구글 자바 컨밴션에 맞춘 구성이네요 👍

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Car {
+
+    private static final int MAXIMUM_CAR_NAME_LENGTH = 5;
```
꼼꼼하고 의미가 명확한 상수화 좋습니다!

sihyung92 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,39 @@
+public class Cars {
+
+    private final List<Car> cars;
+    private final NumberGenerator generator;
```
추상화 관련 리뷰를 반영한다면 Random성으로 인해 테스트가 어려워져 리팩토링해야 하는 수고가 있습니다 ㅎ

ikjo39 replied:
고민 끝에 성장이 있는거겠죠 ㅎ
방법 찾아볼게요!

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,42 @@
+        this.position = DEFAULT_POSITION;
+    }
+
+    public int move(int randomNumber) {
```
NumberGenerator가 주입해주는 숫자이기 때문에 구현에 따라 randomNumber가 들어오지 않을 수도 있으므로 number로 명명짓는게 좋겠어요. 
다른 방향으로 생각해보면 Car구현을 보면 꼭 randomNumber가 들어와야 할 필요가 없는데 Car 객체 외부 로직과 결합도를 높이는 명명입니다

ikjo39 replied:
추상화한 시점에서 다른 가능성을 염두하였는데 변수는 그 의도를 무시하고 확정했단 말씀이시군요.
변수명 또한 그 의도를 고려하여 명명한다.
제 머릿속에 커밋 & 푸시 해둘게요 😁

sihyung92 commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,32 @@
+
+    private void validateTryCount(int tryCount) {
+        if (tryCount < MINIMUM_TRY_COUNT) {
+            throw new IllegalArgumentException();
```
예외 메시지는 가능한 친절하게 작성해주세요! 
```suggestion
            throw new IllegalArgumentException("TryCount가 " + MINIMUM_TRY_COUNT + "보다 작은 숫자가 입력되었습니다. 입력값 : " + tryCount);
```

sihyung92 commented on src/main/java/racingcar/model/RoundResult.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+
+public class RoundResult {
```
이 객체만 봤을 땐 어떤 역할을 하는지 추론하기가 매우 어렵습니다
코드 전체를 읽어야 Integer는 position을 뜻하고 String은 CarName을 뜻하는지 알 수 있습니다.

그리고 또 언제든 규칙에 맞지 않는 값이 들어와 메서드들이 수행하고자 하는 역할을 해칠 수 있습니다
의미를 부여받은 VO들로 Key와 Value 타입을 수정해보는 건 어떨까요?

ikjo39 replied:
제가 봐도 해당 클래스만 봤을 때 추론하기 어려워보이네요. 😅 

> 그리고 또 언제든 규칙에 맞지 않는 값이 들어와 메서드들이 수행하고자 하는 역할을 해칠 수 있습니다

여기서 **규칙에 맞지 않는 값**을 저는 아래처럼 이해하였는데 제가 맞게 이해한걸까요?
(carName, Position)이 아닌 다른 (String, Integer)가 들어왔을 때 구분되지 않는다.

sihyung92 replied:
맞습니다. CarName이라는 것에 대해선 5글자 제약이라던가, 혹은 차이름 이라는 멘탈 모델이 이 객체를 수정하는 개발자에겐 있을거에요. Position도 1~9 사이겠지.. 생각하고 있을텐데요, 

지금의 자료구조는 { "투명드래곤이울부짖엇다00\b\b0123dd" : -2100987 } 같은 값이 들어와도 방어되지 않고, 기존에 가지고 있던 멘탈모델 대로 코딩하다가 버그가 발생할 확률이 높아집니다

ikjo39 replied:
아하! 무슨 말씀이신지 이해되었어요. 감사합니다!

sihyung92 commented on src/main/java/racingcar/model/RoundResult.java:
```diff
@@ -0,0 +1,26 @@
+
+public class RoundResult {
+
+    private LinkedHashMap<String, Integer> result;
```
LinkedHashMap 보다는 Map 인터페이스를 사용하는 건 어떨까요? 이 객체를 활용하는 개발자가 알맞은 컬렉션을 선택하도록 배려할 수 있습니다.

sihyung92 commented on src/main/java/racingcar/model/TotalResult.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public List<String> getWinner() {
+        RoundResult roundResult = totalResult.get(totalResult.size() - 1);
```
lastResult 같은 변수명으로 할당받았다면 가독성이 더 좋았을 것 같습니다.

이 객체에선 생성자에서 인자로 받은 totalResult가 라운드 순서대로 정렬되어있음를 가정하고 있는데요, 이를 검증할 수 있는 코드가 없어 오용되기 쉬운 구조입니다.

TotalResult를 만들 때 빈 List를 가지고 있게 만들고, 
totalResult.add()를 통해 round를 추가하게 했다면 마지막 인덱스가 마지막에 호출된 add()의 RoundResult를 가지고 winner를 선출할 것을 분명히 할 수 있었을 것 같아요

ikjo39 replied:
A1. 변수명도 그 의미가 분명한지 고민해볼게요!
A2. 다시보니 TotalResult 생성시 List로만 매개변수를 받기에 정렬되어있음을 가정할 수 없네요. 반영하겠습니다!

sihyung92 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -0,0 +1,49 @@
+        this.outputView = outputView;
+    }
+
+    public void run() {
+        Cars cars = createCars();
+        RacingGame racingGame = createRacingGame(cars);
+
+        inputView.finishReadingInput();
+
+        TotalResult totalResult = racingGame.run();
+        printResult(totalResult);
+    }
+
+    private Cars createCars() {
+        List<String> carNames = inputView.readCarNames();
+        List<Car> cars = carNames.stream()
+                .map(Car::new)
+                .toList();
+        return new Cars(cars, new RandomNumberGenerator());
+    }
+
+    private RacingGame createRacingGame(Cars cars) {
+        int tryCount = inputView.readTryCount();
+        return new RacingGame(tryCount, cars);
+    }
```
이 리뷰는 의견으로 봐주세요 ㅎ

저는 코드도 문학이라고 봐서요, 10명 중 10명이 다 좋다는 코드는 만들기 힘들다고 생각해요.
그래서 10명 중 7명 정도가 끄덕끄덕 할 정도면 좋은 코드다, 생각하고 또 그렇게 짜려고 노력하는데요.

그런데 제가 생각하기에는, RacingGame이라는 게 있다면 Car 들이 움직이도록 라운드를 진행하고 숫자 같은걸 결정해주는 역할을 할거 같아요. (제가 느끼기에 설득력있는 의견)
그래서 이 부분을 읽으면서 '왜 차들한테 숫자 Generator가 들어가지? RacingGame이 가진게 낫지 않나?' 하는 생각을 했습니다

이건 본인의 생각인 부분이기 때문에 ('내 생각엔 차의 집합에서 숫자를 운영하는게 맞는데?'라고 생각할 수도 있죠 ㅎ 답은 없으니까) 반영하실건 없지만, 좋은 코드라는 건 '많은 사람에게 받아들여지는 코드' 라는 관점이 있다는 부분만 가져가셨으면 좋겠습니다

ikjo39 replied:
> 좋은 코드라는 건 '많은 사람에게 받아들여지는 코드`라는 관점이 있다.

조금 오버를 보태자면 제 가슴을 울리는대요?? 이 부분 또한 고려해볼게요!

sihyung92 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -0,0 +1,39 @@
+                .map(Car::getName)
+                .collect(Collectors.toUnmodifiableSet());
+        if (distinctNames.size() != cars.size()) {
+            throw new IllegalArgumentException();
```
예외메시지를 풍부화해 주세용!

sihyung92 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,25 @@
+        for (RoundResult roundResult : totalResult.getTotalResult()) {
+            LinkedHashMap<String, Integer> result = roundResult.getResult();
+            for (String carName : result.keySet()) {
+                System.out.printf("%s : %s%n", carName, "-".repeat(result.get(carName)));
```
이 부분은 어떤 메시지가 나올지 추론하려면 5초 이상 걸릴 것 같습니다
좀 더 가독성 있는 형식으로 바꿔주세요!

sihyung92 commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,52 @@
+
+class CarTest {
+
+    @DisplayName("자동차 이름에는 null 값이나, 1~5 범위를 벗어난 경우 예외가 발생한다.")
+    @ParameterizedTest
```
다양한 케이스를 실험해볼 수 있는 ParameterizedTest와 꼼꼼한 DisplayedName 좋습니다 ㅎ

sihyung92 commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,52 @@
+        //then
+        assertThat(car.move(given)).isEqualTo(expected);
+    }
+}
```
[EOL](https://avocado12-note.tistory.com/11)이 있네요
이게 문제가 되는 경우는 많지 않지만 - 기호가 보기 싫으니 개행해주도록 합시다

sihyung92 commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @DisplayName("4보다 큰 숫자일 경우 전진, 작을 경우 전진하지 않는다.")
+    @ParameterizedTest
+    @CsvSource(value = {"4,1", "3,0"})
```
조금더 테스트 케이스가 많았으면 좋겠어요. 사실 테스트는 기능명세서 이기도 한데요, 이 두 개의 케이스만 보면 살짝 믿고 쓰기 불안해서요 ㅎ

sihyung92 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,25 @@
+    public void printResult(TotalResult totalResult) {
+        System.out.println();
+        System.out.println("실행 결과");
+        for (RoundResult roundResult : totalResult.getTotalResult()) {
```
Q2. View 에서 비즈니스 로직을 다루지 않는다는 MVC 패턴 규칙을 위배하지 않았나 궁금합니다.

getter를 비즈니스 로직이라고 생각하시나요? getter을 사용하지 않는다면 어떻게 출력 로직을 작성할 수 있을까요? 저는 getter는 비즈니스 로직이 아니라고 생각하는데요, 
비즈니스 로직이란게 뭘지, MVC에서는 왜 View에서 비즈니스 로직을 쓰지 말라고 할지 고민해보시고 알려주셔요 ㅎ

ikjo39 replied:
A1. & A2. 아! 저도 getter는 비즈니스 로직이 아니라 생각해요!
저의 질문 의도는 출력해야하는 최종 문자열까지 (예를들어, 자동차 위치의 경우 "-")  Model에서 변환 후 View로 보내는게 맞지 않을까란 의문이었습니다. 
(급하게 질문을 작성하다보니 명확하게 써놓지 못했네요 :sweat_smile:)

A3. 
그래도 웨지 덕에 초심으로 돌아가 고민해보았어요!
제리의 MVC패턴 테코톡을 참고하여 고민해본 결과는 다음과 같아요.
> 1. 비즈니스 로직이란 데이터를 가공하는 행위이다.
> 2. View는 종류가 많아 데이터를 여러 형태로 보여주어야 하다 보니 비즈니스 로직을 넣을 경우 데이터가 훼손될 가능성이 있다.

이번 미션의 경우 **경주 결과를 "-" 형태로 화면에 출력해주세요.** 까지가 View에서 요구된 요구사항이라 위와 같이 작성하는 것이 맞다는 결론을 내었답니다.

sihyung92 replied:
말씀해주신 부분에 동의합니다 ㅎ

sihyung92 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,42 @@
+
+    private final Scanner SCANNER = new Scanner(System.in);
+
+    public List<String> readCarNames() {
```
Q1. View Layer에서의 역할

토론하셨을 때 다온님 생각은 어떠셨나요? 코드를 보면 List로 분할하는것 까지는 View의 역할로 생각하신 것 같은데 맞나요? ㅎ

MVC는 '도메인은 변화가 적고 View의 변화가 빠르다'는 가정 하에 만들어진 패턴입니다. 
똑같은 도메인 코드를 재활용하고자 하는게 패턴의 목적이니까요,

InputView를 5종류가 있다고 가정하고 코드를 짜보세요.
Json으로 입력되는 View, XML로 전달되는 View, String으로 전달되는 View, 음성인식 View, 미래에 도래할 홀로그램 View..

그러면 View Layer에서 어느정도까지 해줘야 도메인 코드를 안 바꾸고 사용할 수 있을지 고민해볼 수 있을거 같아요.

그리고 패턴은 정답이 아니라는 점도.. 다양하게 경험하시면서 나만의 답을 많이 쌓아두셔요

ikjo39 replied:
저는 처음에는 객체 포장까지 되지 않을까 생각했어요!
그 뒤에 객체 포장 과정에서 비즈니스 로직을 거친다 생각하여 List 분할까지로 의견을 내었습니다 ㅎ

완벽한 정답은 어디에도 없다! 그러나 최선은 있다!
명심하면서 이부분도 같이 고민해나가보겠습니다.

sihyung92 replied:
ㅋㅋ 저는 '정답은 없지만 오답은 있다' 라는 신조가 있습니다

전 View쪽에선 도메인 참조를 해도 된다고 생각해요~ 그러면 도메인 객체 변경에 대한 여파를 View가 맞게 되지만, 그 역의 경우 보단 나아서 InputView, OutputView에서 도메인을 참조(import)하는건 용납하는 편입니다. (리뷰어마다 스타일이 다르겠죠)

sihyung92 reviewed (changes requested):
안녕하세요 다온~~ 리뷰어 웨지입니다!
질문에 대한 답은 리뷰 중에 있습니다~!
리뷰 반영하시고 다시 요청 주세요!

Commit:
refactor: 과도한 추상화 제거

- Car 내 위치 정보를 담을 default 생성자 추가
- 난수 생성 로직 RacingGame 내 이동
- 불분명한 변수명 변경

Commit:
refactor: 적절한 예외 메세지 추가

- 예외 메세지 테스트 코드 추가

Commit:
feat: VO 추가

- 자동차 이름
- 차 위치

Commit:
refactor: RoundResult 내 (Key, Value) 타입 변경

- 불분명한 타입 -> VO로 변경
- (String, Integer) -> (CarName, Position)
- LinkedHashMap -> Map 인터페이스 변경

Commit:
refactor: 우승자 선택 리팩토링

- totalResult.add() 로직 추가
- RoundResult lastResult 필드 추가

Commit:
refactor: 가독성 향상을 위한 메서드 분리

- 경주 결과 출력 메서드 분리

Commit:
test: carMovingTest 내 테스트 케이스 추가

Commit:
style: 불필요한 줄바꿈 제거 및 EOL 추가

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -32,7 +38,8 @@ public int getPosition() {
     private void validateName(String name) {
         if (isNameEmpty(name) || name.length() > MAXIMUM_CAR_NAME_LENGTH) {
-            throw new IllegalArgumentException();
+            String message = "자동차 이름이 1-5자 사이의 길이로 입력되지 않았습니다.";
```
상세한 예외메시지 좋아요 ㅎ 조금더 변화에 잘 대응되며 풍부한 정보를 담기 위해 아래와 같이도 가능할거같네요
```suggestion
            String message = "자동차 이름이 1-" + MAXIMUM_CAR_NAME_LENGTH + "자 사이의 길이로 입력되지 않았습니다. 입력값 : " + name;
```

ikjo39 replied:
입력값을 보여주며 사용자 입장에서 더 친절한 예외 메세지가 되겠네요 ㅎ
반영해볼게요!

sihyung92 commented on src/main/java/racingcar/model/CarName.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
```
VO에 대해 꼼꼼한 equals & hashcode 오버라이드 👍
기왕 하는김에 toString도 해두면 낫습니다 ㅎ 
toString 오버라이딩 해놓는 이유는 보통 디버깅 상황에서 도움을 받기 위함입니다 (프로덕션 코드에서 활용하기엔 적절하지 않음)

ikjo39 replied:
> (프로덕션 코드에서 활용하기엔 적절하지 않음)

세세한 설명까지... 감사합니다.

sihyung92 commented on src/main/java/racingcar/model/Position.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.Objects;
+
+public class Position {
```
원시값 맵핑 좋아용 ㅎ

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -15,8 +15,14 @@ public Car(String name) {
-    public int move(int randomNumber) {
-        if (randomNumber >= MINIMUM_STEP_FORWARD_STANDARD) {
+    Car(String name, int position) {
+        validateName(name);
+        this.name = name;
+        this.position = position;
```
이 객체들도 새롭게 생성해주신 CarName와 Position 객체를 활용할 수 있지 않을까요?
validateName의 코드를 CarName 생성자로 옮겨주시면 CarName을 활용하는 모든 곳에서 동일한 validation을 보장받을 수 있습니다

ikjo39 replied:
VO 객체 내에 Validation 로직이 들어가도 될까? 란 고민이 시작되는데요. (제가 호기심 대마왕이긴 합니다 ㅎ)
이 부분 함께 고민하여 다음 스텝에 반영해보겠습니다 😁

sihyung92 commented on src/main/java/racingcar/model/RoundResult.java:
```diff
@@ -1,26 +1,35 @@
     }
+
+    private int getMaxPosition() {
+        List<Integer> positions = result.values()
+                .stream()
+                .map(Position::getPosition)
+                .toList();
+        return Collections.max(positions);
```
다른 스트림 연산을 사용하면 이렇게도 가능할거 같네요 
```suggestion
      return result.values()
          .stream()
          .mapToInt(Position::getPosition)
          .max()
          .orElse(0);
```

ikjo39 replied:
Position 객체 내 position 이 int 형이니 `.mapToInt()` 를 바로 거쳐도 되는군요.
두 기능을 하나로 합치니 더 직관적이게 보이네요!

sihyung92 commented on src/main/java/racingcar/model/TotalResult.java:
```diff
@@ -1,17 +1,19 @@
 public class TotalResult {
-    private final List<RoundResult> totalResult;
+    private final List<RoundResult> totalResult = new ArrayList<>();
+    private RoundResult lastResult;
```
캐싱해두는 아이디어가 괜찮네요 ㅎ

sihyung92 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -28,15 +28,17 @@ public void finishReadingInput() {
     private static void validateInput(String input) {
         if (input == null || input.isEmpty() || input.endsWith(",")) {
-            throw new IllegalArgumentException();
+            String message = "자동차 이름 입력은 공백이거나 구분자(,)로 끝날 수 없습니다.";
```
해당 클래스의 오류도 input을 오류 메시지에 포함해 풍부화 해주면 더 좋을거 같습니당!

ikjo39 replied:
헉 빠뜨린 부분이네요! 반영해보겠습니다.

sihyung92 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -10,16 +12,25 @@ public class OutputView {
+    private void convertResultText(CarName carName, Map<CarName, Position> result) {
+        String name = carName.getName();
+        int position = result.get(carName).getPosition();
+        System.out.printf("%s : %s%n", name, "-".repeat(position));
```
저번 리뷰에선 형식 파악에 몇 초 걸릴거 같다고 했는데 다시보니 괜찮은거 같기도 하고요?

그래도 저는 이런 케이스에선 주석을 적극적으로 활용합니다 (다른 개발자가 이 형식을 단번에 파악할 수 있을까? 고민해보고 아니라면 주석 사용)
클린코드나 우테코에선 주석 대신 코드로 표현하라고 하지만, 왜 쓰지말라고 하는지 파악해두고 필요할 땐 사용하는 게 좋다고 생각합니다 ㅎ
```suggestion
        // ex) Pobi : ------ 
        System.out.printf("%s : %s%n", name, "-".repeat(position));
```

ikjo39 replied:
> 다른 개발자가 이 형식을 단번에 파악할 수 있는지 고민해보고 아니라면 과감히 주석 사용
> 하지만, 왜 쓰지 말라고 하는지 파악해두자

이게 가장 중요한 부분이네요! 한번 정리해보겠습니다 ㅎ

sihyung92 reviewed (approved):
안녕하세요 다온~ 
리뷰를 잘 반영해주셨네요 ㅎ
추가적인 리뷰가 있는데 이 내용들은 step2에서 반영해주세요.
감사합니다~!
